orchagent/debug_counter/debug_counter.cpp

```diff
@@ -0,0 +1,99 @@
+#include "debug_counter.h"
+#include "drop_counter.h"
+
+#include <unordered_set>
+#include <unordered_map>
+#include <string>
+#include <memory>
+#include <vector>
+#include "rediscommand.h"
+#include <exception>
+#include "logger.h"
+
+using std::runtime_error;
+using std::string;
+using std::unordered_map;
+using std::unique_ptr;
+using std::unordered_set;
+using std::vector;
+using swss::FieldValueTuple;
+
+extern sai_object_id_t gSwitchId;
+extern sai_debug_counter_api_t *sai_debug_counter_api;
+
+// It is expected that derived types populate any relevant fields and
+// initialize the counter in the SAI.
+//
+// If counter_type is not a member of debug_counter_type_lookup then this
+// constructor will throw a runtime error.
+DebugCounter::DebugCounter(
+        const string& counter_name,
+        const string& counter_type)
+    : name(counter_name)
+{
+    SWSS_LOG_ENTER();
+
+    auto counter_type_it = debug_counter_type_lookup.find(counter_type);
+    if (counter_type_it == debug_counter_type_lookup.end()) {
+        SWSS_LOG_ERROR("Failed to initialize debug counter of type '%s'",
+                counter_type.c_str());
+        throw runtime_error("Failed to initialize debug counter");
+    }
+    this->type = counter_type_it->first;
+}
+
+// It is expected that derived types delete the counter from the SAI.
+DebugCounter::~DebugCounter()
+{
+    SWSS_LOG_ENTER();
+}
+
+// serializeDebugCounterMetadata populates ONLY the type field in
+// serialized_attributes.
+//
+// Behavior is undefined if the length of serialized_attributes is less than 1.
+void DebugCounter::serializeDebugCounterMetadata(sai_attribute_t *serialized_attributes)
```

Comment:
Why not use sai_attribute_t& serialized_attribute for serialization since you are only putting values in the first entry anyways?